test/fennec/auth_test.exs

```diff
@@ -0,0 +1,303 @@
+  end
+
+  test "empty request returns nonce and realm", ctx do
+    udp = ctx.udp
+    id = :crypto.strong_rand_bytes(12)
+    req =
+      allocate_request(id)
+      |> Format.encode()
+
+    resp = udp_communicate(udp, 0, req)
+
+    params = Format.decode!(resp)
+    assert %Params{class: :failure,
+                   method: :allocate,
+                   identifier: ^id} = params
+
+    assert %ErrorCode{code: 401} = Params.get_attr(params, ErrorCode)
+    assert %Nonce{value: nonce} = Params.get_attr(params, Nonce)
+    assert %Realm{value: realm} = Params.get_attr(params, Realm)
+
+    assert String.length(nonce) > 0
+    assert String.length(nonce) < 128
+    assert byte_size(nonce) < 763
+
+    assert String.length(realm) > 0
+    assert String.length(realm) < 128
+    assert byte_size(realm) < 763
+  end
+
+  test "request with all missing attributes fails to authenticate", ctx do
+    udp = ctx.udp
+    id = :crypto.strong_rand_bytes(12)
+    req =
+      allocate_request(id)
+      |> Format.encode(secret: "abc", realm: "realm", username: "user")
+
+    resp = udp_communicate(udp, 0, req)
+
+    params = Format.decode!(resp)
+    assert %Params{class: :failure,
+                   method: :allocate,
+                   identifier: ^id} = params
+
+    assert %ErrorCode{code: 400} = Params.get_attr(params, ErrorCode)
+  end
+
+  test "request with missing nonce attribute fails to authenticate", ctx do
+    udp = ctx.udp
+    id = :crypto.strong_rand_bytes(12)
+    attrs = [
+      %RequestedTransport{protocol: :udp},
+      %Realm{value: "localhost"},
+      %Username{value: "user"}
+    ]
+    req =
+      allocate_request(id, attrs)
+      |> Format.encode(secret: "abc")
+
+    resp = udp_communicate(udp, 0, req)
+
+    params = Format.decode!(resp)
+    assert %Params{class: :failure,
+                   method: :allocate,
+                   identifier: ^id} = params
+
+    assert %ErrorCode{code: 400} = Params.get_attr(params, ErrorCode)
+  end
+
+  test "request with missing username attributes fails to authenticate", ctx do
+    udp = ctx.udp
+    id = :crypto.strong_rand_bytes(12)
+    attrs = [
+      %RequestedTransport{protocol: :udp},
+      %Realm{value: "localhost"},
+      %Nonce{value: "nonce"}
+    ]
+    req =
+      allocate_request(id, attrs)
+      |> Format.encode(secret: "abc", username: "user")
+
+    resp = udp_communicate(udp, 0, req)
+
+    params = Format.decode!(resp)
+    assert %Params{class: :failure,
+                   method: :allocate,
+                   identifier: ^id} = params
+
+    assert %ErrorCode{code: 400} = Params.get_attr(params, ErrorCode)
+  end
+
+  test "request with missing realm attributes fails to authenticate", ctx do
+    udp = ctx.udp
+    id = :crypto.strong_rand_bytes(12)
+    attrs = [
+      %RequestedTransport{protocol: :udp},
+      %Username{value: "user"}
+    ]
+    req =
+      allocate_request(id, attrs)
+      |> Format.encode(secret: "abc", realm: "localhost")
```

Comment:
Looking at this I don't know if secret passed to `encode/2` is valid or not. I'd put valid and invalid secrets in module attributes.